graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlCache.java

```diff
@@ -232,70 +220,99 @@ public void onKeyspaceChanged(
+     * Note that the source can be {@code null} to indicate that no custom schema was deployed.
+     */
+    boolean matches(@Nullable SchemaSource source);
 
-    DmlGraphqlHolder(SchemaSource source, Keyspace keyspace) {
-      this.source = source;
-      this.keyspace = keyspace;
-    }
+    boolean isCqlFirst();
+  }
 
-    DmlGraphqlHolder(SchemaSource source, GraphQL graphQL) {
+  /**
+   * An entry used when we need to recompute the schema (because we're reacting to the insertion of
+   * a new row in the source table).
+   */
+  class LazyGraphqlHolder implements GraphqlHolder {
+    private final SchemaSource source;
+    private final Supplier<GraphQL> computeGraphql;
+    private volatile GraphQL graphql;
+
+    LazyGraphqlHolder(SchemaSource source, Keyspace keyspace) {
       this.source = source;
-      this.graphqlFuture.complete(graphQL);
-      // when the graphqlFuture is already completed, the keyspace is not-needed
-      this.keyspace = null;
+      this.computeGraphql =
+          (source == null)
+              ? () -> newGraphql(SchemaFactory.newDmlSchema(keyspace))
+              : () -> {
+                ProcessedSchema processedSchema =
+                    new SchemaProcessor(persistence, true).process(source.getContents(), keyspace);
+                // Check that the data model still matches
+                CassandraMigrator.forPersisted()
+                    .compute(processedSchema.getMappingModel(), keyspace);
+                return processedSchema.getGraphql();
+              };
     }
 
-    void init() {
-      try {
-        graphqlFuture.complete(buildGraphql());
-      } catch (Exception e) {
-        graphqlFuture.completeExceptionally(e);
+    @Override
+    public GraphQL getGraphql() {
+      GraphQL result = graphql;
+      if (result != null) {
+        return result;
       }
-    }
-
-    private GraphQL buildGraphql() {
-
-      if (source == null) {
-        return newGraphql(SchemaFactory.newDmlSchema(keyspace));
-      } else {
-        ProcessedSchema processedSchema =
-            new SchemaProcessor(persistence, true).process(source.getContents(), keyspace);
-        // Check that the data model still matches
-        CassandraMigrator.forPersisted().compute(processedSchema.getMappingModel(), keyspace);
-
-        return processedSchema.getGraphql();
+      synchronized (this) {
+        if (graphql == null) {
+          graphql = computeGraphql.get();
+        }
+        return graphql;
       }
     }
 
-    boolean isSameVersionAs(SchemaSource otherSource) {
+    @Override
+    public boolean matches(@Nullable SchemaSource otherSource) {
       return (source == null && otherSource == null)
           || (source != null
               && otherSource != null
               && source.getVersion().equals(otherSource.getVersion()));
     }
 
-    boolean isCqlFirst() {
+    @Override
+    public boolean isCqlFirst() {
       return source == null;
```

Comment:
I don't fully understand why it is cqlFirst in that case, can you elaborate more?

Reply:
When there is no entry in the `schema_source` table, we expose the legacy schema.